include/sparrow_v01/layout/primitive_array.hpp

```diff
@@ -0,0 +1,187 @@
+        using inner_types = array_inner_types<self_type>;
+        using inner_value_type = typename inner_types::inner_value_type;
+        using inner_reference = typename inner_types::inner_reference;
+        using inner_const_reference = typename inner_types::inner_const_reference;
+        using bitmap_type = typename base_type::bitmap_type;
+        using bitmap_reference = typename base_type::bitmap_reference;
+        using bitmap_const_reference = typename base_type::bitmap_const_reference;
+        using value_type = nullable<inner_value_type>;
+        using reference = nullable<inner_reference, bitmap_reference>;
+        using const_reference = nullable<inner_const_reference, bitmap_const_reference>;
+        using pointer = typename inner_types::pointer;
+        using const_pointer = typename inner_types::const_pointer;
+        using size_type = typename base_type::size_type;
+        using difference_type = typename base_type::difference_type;
+        using iterator_tag = std::contiguous_iterator_tag;
+
+        using value_iterator = typename base_type::value_iterator;
+        using const_value_iterator = typename base_type::const_value_iterator;
+
+        explicit primitive_array(arrow_proxy);
+        virtual ~primitive_array() = default;
+
+        using base_type::size;
+
+        reference operator[](size_type i);
+        const_reference operator[](size_type i) const;
+
+    private:
+
+        using base_type::storage;
+        using base_type::has_value;
+        using base_type::bitmap_begin;
+        using base_type::bitmap_end;
+
+        pointer data();
+        const_pointer data() const;
+
+        inner_reference value(size_type i);
+        inner_const_reference value(size_type i) const;
+
+        value_iterator value_begin();
+        value_iterator value_end();
+
+        const_value_iterator value_cbegin() const;
+        const_value_iterator value_cend() const;
+
+        primitive_array(const primitive_array&) = default;
+        primitive_array* clone_impl() const override;
+
+        friend class array_crtp_base<self_type>;
+    };
+
+    /**********************************
+     * primitive_array implementation *
+     **********************************/
+
+    template <class T>
+    primitive_array<T>::primitive_array(arrow_proxy proxy)
+        : array_base()
+        , base_type(std::move(proxy))
+    {
+    }
+
+    template <class T>
+    auto primitive_array<T>::operator[](size_type i) -> reference
+    {
+        SPARROW_ASSERT_TRUE(i < size());
+        return reference(value(i), has_value(i));
+    }
+
+    template <class T>
+    auto primitive_array<T>::operator[](size_type i) const -> const_reference
+    {
+        SPARROW_ASSERT_TRUE(i < size());
+        return reference(value(i), has_value(i));
+    }
+
+    template <class T>
+    auto primitive_array<T>::data() -> pointer
+    {
+        SPARROW_ASSERT_TRUE(storage().n_buffers() > 1);
```

Comment:
This can be removed if you check that the data_type is correct in the constructors